travis-ci/packer-templates-mac
# Make rake tasks run on image-builder automatically

## What is the problem that this PR is trying to fix?

It's confusing that we use a Rakefile to build, but for building images and checking out hosts, you need to use the `bin/build-image` script to actually have the tools set up to run your task correctly.

## What approach did you choose and why?

This change defines a `remote_task` DSL method for defining Rake tasks that should only run on the image-builder. When run on a machine other than the image-builder, the task will automatically be run on the image-builder for you, using the existing `bin/build-image` script, which has now been renamed to `bin/rake-remote`.

Part of the motivation for doing this is that I want to be able to put some hooks around image builds that need to run on the dev machine, not the image-builder. Previously, that couldn't be done with Rake, because Rake was only being run on the image-builder, not the dev machine. Now Rake is a part of the process on both machines, which enables extending those tasks with the hooks needed.

For instance, we can add some work on the dev machine after the `xcode10.0` image builds by extending the local variant of the task. In Rake, if you redefine an existing task, it will append to the work that task does:

```
task 'xcode10.0:local' do
  # sync some data back from the image-builder
end
```

Similarly, we can add work before the task runs by adding a new prerequisite to the local task.

## How can you test this?

Run `rake host:checkout`, `rake host:checkin`, or build an image. They should run rake remotely on the image-builder, which should do the appropriate action.

## What feedback would you like, if any?

Any.
